memory/iter.go

```diff
@@ -7,210 +7,174 @@ import (
 	"github.com/tidwall/btree"
 )
 
-type dsPrefixIter struct {
-	ctx     context.Context
+type iterator struct {
 	db      *Datastore
 	version uint64
-	it      *baseIterator
-	prefix  []byte
-	reverse bool
-	curItem *dsItem
-}
+	it      btree.IterG[dsItem]
 
-type dsRangeIter struct {
-	ctx     context.Context
-	db      *Datastore
-	version uint64
-	it      *baseIterator
-	start   []byte
-	end     []byte
-	reverse bool
-	curItem *dsItem
-}
+	// The key at which this iterator begins.
+	//
+	// This is inclusive only if `mayExactlyMatchStart` is true.
+	start []byte
 
-func newPrefixIter(ctx context.Context, db *Datastore, prefix []byte, reverse bool, version uint64) *dsPrefixIter {
-	pIter := &dsPrefixIter{
-		ctx:     ctx,
-		db:      db,
-		version: version,
-		it:      newBaseIterator(db.values, reverse, prefix, bytesPrefixEnd(prefix)),
-		prefix:  prefix,
-		reverse: reverse,
-	}
-	if reverse {
-		pIter.Seek(bytesPrefixEnd(prefix))
-	} else {
-		pIter.Seek(prefix)
-	}
+	// The key at which this iterator ends, inclusive.
+	end []byte
 
-	// if the first key is an exact match to the prefix, skip next
-	// since prefix is a *strict* subset prefix
-	if pIter.curItem != nil && bytes.Equal(pIter.Key(), prefix) {
-		pIter.Next()
-	}
+	// If this is true, `start` is inclusive, else `start` is exclusive.
+	mayExactlyMatchStart bool
```

Comment:
nitpick: when `mayExactlyMatchStart` is true, `start` is for sure inclusive it seems like which is why the `may` might be confusing. Suggesting renaming to perhaps, `isStartInclusive` or `isStrictStart`.

Reply:
Good shout, thanks Shahzad - will change.

- [ ] Rename `mayExactlyMatchStart`